pages/en/zkapps/advanced-concepts.mdx

```diff
@@ -0,0 +1,64 @@
+import Page from "@reason/pages/Docs";
+export default Page({
+  title: "Advanced Concepts | Mina Protocol",
+  description: "Check out these advanced concepts regarding zkApps.",
+});
+
+# Advanced Concepts
+
+Build onto what you already learned regarding zkApps and SnarkyJS
+
+This page explores additional topics that requires some familiarity with building zkApps, SnarkyJS, and crytography.
+
+## Events
+
+Events represent arbitrary information that can be passed along with a transaction.
+
+An example use case are zkApps which keep some large internal state, and only store a commitment to that internal state on-chain. For example, a Merkle tree where only the root is stored in on-chain state. Events enable to attach the full information of state changes in transactions. In the Merkle tree example, this could mean sending any Merkle leaves that are changed by the transaction as events. This means that an observer of these transactions can follow along and keep track of the full Merkle tree on their side.
+
+To use events, you have to declare an `events` field at the top level of your smart contract. It contains the _names_ and _types_ of your events. Here's an example:
+
+```ts
+class MyContract extends SmartContract {
+  events = {
+    "add-merkle-leaf": Field,
+    "update-merkle-leaf": Field,
+  };
+}
+```
+
+In this example, we declare events called `"add-merkle-leaf"` and `"update-merkle-leaf"`, both with a type of `Field`. Instead of `Field`, you can also use other built-in snarkyjs types as well as any `CircuitValue`. (In fact, a custom `CircuitValue` is probably better-suited to encode leaves of a Merkle tree -- we just use `Field` for simplicity here.)
+
+After declaring your events, you can use `this.emitEvent(name, event)` in any smart contract method, where `event` has to have the type you declared for that `name`. Example:
+
+```ts
+class MyContract extends SmartContract {
+  events = {
+    "add-merkle-leaf": Field,
+    "update-merkle-leaf": Field,
+  }
+
+  @method updateMerkleTree(leaf: Field, ...) {
+    this.emitEvent("update-merkle-leaf", leaf);
+    // ...
+  }
+}
+```
+
+Some other important facts about events:
+
+- Events are not stored on-chain. Only events from the last couple of transactions will be kept by consensus nodes; after that, they will be discarded.
```

Comment:
> Events are not stored on-chain. Only events from the most recent couple of transactions are retained by consensus nodes. After that, they will be discarded, but are still retained on archive nodes. In the near future, we plan to add an API to easily fetch events for any smart contract from an archive node.

Would suggest merging bullets 1 & 2 to ensure nobody gets a wrong understanding after reading the first, especially if they're just skimming the doc.